### PR TITLE
Add a new status: changes requested

### PR DESCRIPTION
**Pull request by amieiro**
## Problem

This PR, with [another one](https://github.com/GlotPress/gp-translation-helpers/pull/65) in the [gp-translation-helpers](https://github.com/GlotPress/gp-translation-helpers) plugin, adds a new status: changes requested.

## Solution

GlotPress stores a new status (changes_requested) when:
- A simple rejection is made, adding some feedback.
- A bulk rejection is made, adding some feedback.

## Testing instructions

To test this PR, you have to use the [related branch](https://github.com/GlotPress/gp-translation-helpers/pull/65) from [gp-translation-helpers](https://github.com/GlotPress/gp-translation-helpers) plugin and you have to reject some strings as validator (simple or bulk rejection), adding some feedback.

## Screenshots or screencast 

The new status in the main table:

image

The new status in a string:

image

Rejecting a string:

image

When a validator adds feedback (reason and/or comment), the "Reject" button becomes "Request changes", so the validator sets the string in this new status. A notification is sent to the translator.

image

Making a bulk reject:

image

When a validator adds feedback (reason and/or comment), the "Reject" button becomes "Request changes", so the validator sets the string in this new status. A notification is sent to the translator.

image

The new status in the filters:

image

### CODE REVIEW

File: gp-templates/helper-functions.php [hunk 1]
```diff
@@ -366,12 +366,16 @@ function textareas( $entry, $permissions, $index = 0 ) {
...
-		'rejected' => _x( 'rejected', 'Single Status', 'glotpress' ),
+		'current'           => _x( 'current', 'Single Status', 'glotpress' ),
+		'waiting'           => _x( 'waiting', 'Single Status', 'glotpress' ),
+		'fuzzy'             => _x( 'fuzzy', 'Single Status', 'glotpress' ),
+		'old'               => _x( 'old', 'Single Status', 'glotpress' ),
+		'rejected'          => _x( 'rejected', 'Single Status', 'glotpress' ),
+		'changes_requested' => _x( 'Changes requested', 'Single Status', 'glotpress' ),
 	);
+	if ( ! apply_filters( 'gp_translation_helper_installed', false ) ) {// todo: delete when we merge the gp-translation-helpers in GlotPress
+		$status_labels['changes_requested'] = _x( 'rejected', 'Single Status', 'glotpress' );
```
Comment (akirk) [inline gp-templates/helper-functions.php]: Could you add a comment to explain the scenario? If a `changes_requested` status exists in the database but they are no longer enabled, they will show as rejected.

File: gp-templates/translation-row-editor-meta-status.php [hunk 2]
```diff
@@ -11,20 +11,25 @@
...
+					<?php endif; ?>
+					<?php if ( ( 'rejected' !== $translation->translation_status ) && ( 'changes_requested' !== $translation->translation_status ) ) : ?>
+						<button class="button is-small reject" data-nonce="<?php echo esc_attr( wp_create_nonce( 'update-translation-status-rejected_' . $translation->id ) ); ?>" title="<?php esc_attr_e( 'Reject this translation. The existing translation will be kept as part of the translation history.', 'glotpress' ); ?>"><strong>&minus;</strong> <?php _ex( 'Reject', 'Action', 'glotpress' ); ?></button>
+						<?php if ( apply_filters( 'gp_enable_changes_requested_status', false ) ) : // todo: delete when we merge the gp-translation-helpers in GlotPress ?>
+							<button class="button is-small changes_requested" style="display: none;" data-nonce="<?php echo esc_attr( wp_create_nonce( 'update-translation-status-changes_requested_' . $translation->id ) ); ?>" title="<?php esc_attr_e( 'Request changes for this translation. The existing translation will be kept as part of the translation history.', 'glotpress' ); ?>"><strong>&minus;</strong> <?php _ex( 'Request changes', 'Action', 'glotpress' ); ?></button>
```
Comment (akirk) [inline gp-templates/translation-row-editor-meta-status.php]: I think we should not use underscores in css classes for consistency reasons.

File: gp-templates/translation-row-editor-meta-status.php [hunk 3]
```diff
@@ -11,20 +11,25 @@
...
-					<button class="button is-small approve" data-nonce="<?php echo esc_attr( wp_create_nonce( 'update-translation-status-current_' . $translation->id ) ); ?>" title="<?php esc_attr_e( 'Approve this translation. Any existing translation will be kept as part of the translation history.', 'glotpress' ); ?>"><strong>+</strong> <span><?php _ex( 'Approve', 'Action', 'glotpress' ); ?></span></button>
-				<?php endif; ?>
-				<?php if ( 'rejected' !== $translation->translation_status ) : ?>
-					<button class="button is-small reject" data-nonce="<?php echo esc_attr( wp_create_nonce( 'update-translation-status-rejected_' . $translation->id ) ); ?>" title="<?php esc_attr_e( 'Reject this translation. The existing translation will be kept as part of the translation history.', 'glotpress' ); ?>"><strong>&minus;</strong> <?php _ex( 'Reject', 'Action', 'glotpress' ); ?></button>
-				<?php endif; ?>
-				<?php if ( 'fuzzy' !== $translation->translation_status ) : ?>
+		<?php if ( ! ( ( 'changes_requested' == $translation->translation_status ) && ( ! apply_filters( 'gp_enable_changes_requested_status', false ) ) ) ) : // todo: delete when we merge the gp-translation-helpers in GlotPress. ?>
```
Comment (akirk) [inline gp-templates/translation-row-editor-meta-status.php]: This `if` is confusing. Can we make it easier? Both remove the double negation and parenthesis.

File: gp-templates/helper-functions.php [hunk 4]
```diff
@@ -366,12 +369,17 @@ function textareas( $entry, $permissions, $index = 0 ) {
...
-		'waiting'  => _x( 'waiting', 'Single Status', 'glotpress' ),
-		'fuzzy'    => _x( 'fuzzy', 'Single Status', 'glotpress' ),
-		'old'      => _x( 'old', 'Single Status', 'glotpress' ),
-		'rejected' => _x( 'rejected', 'Single Status', 'glotpress' ),
+		'current'           => _x( 'current', 'Single Status', 'glotpress' ),
+		'waiting'           => _x( 'waiting', 'Single Status', 'glotpress' ),
+		'fuzzy'             => _x( 'fuzzy', 'Single Status', 'glotpress' ),
+		'old'               => _x( 'old', 'Single Status', 'glotpress' ),
+		'rejected'          => _x( 'rejected', 'Single Status', 'glotpress' ),
+		'changes_requested' => _x( 'Changes requested', 'Single Status', 'glotpress' ),
```
Comment (akirk) [inline gp-templates/helper-functions.php]: All the other statuses start with lower case. This one should too, I think.

File: gp-templates/helper-functions.php [hunk 5]
```diff
@@ -366,12 +369,17 @@ function textareas( $entry, $permissions, $index = 0 ) {
...
 	);
+	// If a changes_requested status exists in the database but they are no longer enabled, they will show as rejected.
+	if ( ! apply_filters( 'gp_enable_changes_requested_status', false ) ) {// todo: delete when we merge the gp-translation-helpers in GlotPress
+		$status_labels['changes_requested'] = _x( 'untranslated', 'Single Status', 'glotpress' );
```
Comment (akirk) [inline gp-templates/helper-functions.php]: Comment and code are not consistent here. untranslated or rejected?

File: gp-templates/helper-functions.php [hunk 6]
```diff
@@ -341,6 +341,9 @@ function textareas( $entry, $permissions, $index = 0 ) {
 			endforeach;
 
 		endif;
+		if ( ( 'changes_requested' == $entry->translation_status ) && ( ! apply_filters( 'gp_enable_changes_requested_status', false ) ) ) { // todo: delete when we merge the gp-translation-helpers in GlotPress
```
Comment (akirk) [inline gp-templates/helper-functions.php]: The inner parenthesis are unnecessary on both clauses.

File: gp-templates/helper-functions.php [hunk 7]
```diff
@@ -341,6 +341,9 @@ function textareas( $entry, $permissions, $index = 0 ) {
...
+			$entry->translations = array();
```
Comment (akirk) [inline gp-templates/helper-functions.php:347]: Could you add a comment here why this is necessary?

**Review by akirk — COMMENTED** [review]
I reviewed this from a code perspective and something about the underscore makes it seem off. Could we rename `changes_requested` to something without underscore or hyphen. Maybe just`changesrequested`. Or `changerequest` or `reviewed`?

File: gp-templates/translation-row-editor-meta-status.php (hunk 2 above)
Comment (amieiro) [reply gp-templates/translation-row-editor-meta-status.php]: I have updated these texts: `changes_requested` to `changesrequested` and `request_changes` to `requestchanges`. Done here https://github.com/GlotPress/GlotPress/pull/1451/commits/da177763d4f85689b0d636238eec37e1a2167bb5.

File: gp-templates/helper-functions.php (hunk 4 above)
Comment (amieiro) [reply gp-templates/helper-functions.php]: Updated here https://github.com/GlotPress/GlotPress/pull/1451/commits/025e9330e97181bb1b8aef3e272a43416ca6ff23.

File: gp-templates/helper-functions.php (hunk 5 above)
Comment (amieiro) [reply gp-templates/helper-functions.php]: Good catch! The status was incorrect. Solved here https://github.com/GlotPress/GlotPress/pull/1451/commits/5e286f3db15f581ac68cd0023a7a0513912b8be8.

File: gp-templates/translation-row-editor-meta-status.php (hunk 3 above)
Comment (amieiro) [reply gp-templates/translation-row-editor-meta-status.php]: Done here https://github.com/GlotPress/GlotPress/pull/1451/commits/8dc5c3249c49e6a232d32d0fdee74d06098a5546.

File: gp-templates/helper-functions.php (hunk 7 above)
Comment (amieiro) [reply gp-templates/helper-functions.php:347]: Done here https://github.com/GlotPress/GlotPress/pull/1451/commits/d534453390a58ab36e4af669aea9716fe97082ac.

File: assets/js/editor.js [hunk 8]
```diff
@@ -106,6 +106,7 @@ $gp.editor = (
 					.on( 'click', 'button.insertnl', $gp.editor.hooks.newline )
 					.on( 'click', 'button.approve', $gp.editor.hooks.set_status_current )
 					.on( 'click', 'button.reject', $gp.editor.hooks.set_status_rejected )
+					.on( 'click', 'button.changesrequested', $gp.editor.hooks.set_status_changesrequested )
```
Comment (akirk) [inline assets/js/editor.js:109]: If the `gp-translation-helpers` plugin is not active or not on the https://github.com/GlotPress/gp-translation-helpers/pull/65 brnach, then this causes this error message and breaks most of the JS on the page:
```
Uncaught TypeError: can't assign to property "guid" on "button.changesrequested": not an object
```

Comment (akirk) [reply assets/js/editor.js:109]: A fix would be to add a stub like this below:
```
				set_status_changesrequested: function() {
					return false;
				},
```

**Review by akirk — CHANGES_REQUESTED** [review]
The JS error needs to be fixed because it breaks much of the GlotPress UI.

File: assets/js/editor.js (hunk 8 above)
Comment (amieiro) [reply assets/js/editor.js:109]: Good catch. Resolved [here](https://github.com/GlotPress/GlotPress/pull/1451/commits/65147031bd25e43f2c15a24e51887565fd252a96).

**Review by akirk — APPROVED** [review]
I've tested that the PR doesn't change the behavior of GlotPress unless I use it in combination with https://github.com/GlotPress/gp-translation-helpers/pull/65, which is exactly what we want.

**Pull request merged**